### PR TITLE
Improve performance of equals and equalsUnchecked

### PR DESCRIPTION
**Pull request by fmeiser**
This PR improves performance of `equals `and `equalsUnchecked`. This is accomplished by using address directly with unsafe.

Before `equals`
```
Benchmark                       (size)  Mode  Cnt     Score   Error  Units
BenchmarkSlice.equalsUnchecked       1  avgt   50     4,694 ± 0,065  ns/op
BenchmarkSlice.equalsUnchecked       7  avgt   50    15,023 ± 0,088  ns/op
BenchmarkSlice.equalsUnchecked       8  avgt   50     4,857 ± 0,041  ns/op
BenchmarkSlice.equalsUnchecked      16  avgt   50     8,303 ± 0,060  ns/op
BenchmarkSlice.equalsUnchecked      32  avgt   50    11,382 ± 0,051  ns/op
BenchmarkSlice.equalsUnchecked      64  avgt   50    17,036 ± 0,513  ns/op
BenchmarkSlice.equalsUnchecked     127  avgt   50    36,838 ± 0,061  ns/op
BenchmarkSlice.equalsUnchecked   32779  avgt   50  5674,848 ± 6,782  ns/op
```

After equals
```
Benchmark                       (size)  Mode  Cnt     Score    Error  Units
BenchmarkSlice.equalsUnchecked       1  avgt   50     4,854 ±  0,055  ns/op
BenchmarkSlice.equalsUnchecked       7  avgt   50    11,554 ±  0,161  ns/op
BenchmarkSlice.equalsUnchecked       8  avgt   50     4,911 ±  0,060  ns/op
BenchmarkSlice.equalsUnchecked      16  avgt   50     6,143 ±  0,112  ns/op
BenchmarkSlice.equalsUnchecked      32  avgt   50     9,221 ±  0,112  ns/op
BenchmarkSlice.equalsUnchecked      64  avgt   50    14,365 ±  0,123  ns/op
BenchmarkSlice.equalsUnchecked     127  avgt   50    26,765 ±  0,191  ns/op
BenchmarkSlice.equalsUnchecked   32779  avgt   50  3844,702 ± 12,107  ns/op
```

Before `equalsObject`
```
Benchmark                    (size)  Mode  Cnt     Score   Error  Units
BenchmarkSlice.equalsObject       1  avgt   50     4,810 ± 0,035  ns/op
BenchmarkSlice.equalsObject       7  avgt   50    16,664 ± 0,079  ns/op
BenchmarkSlice.equalsObject       8  avgt   50     5,093 ± 0,075  ns/op
BenchmarkSlice.equalsObject      16  avgt   50     8,578 ± 0,045  ns/op
BenchmarkSlice.equalsObject      32  avgt   50    12,319 ± 0,072  ns/op
BenchmarkSlice.equalsObject      64  avgt   50    18,464 ± 0,052  ns/op
BenchmarkSlice.equalsObject     127  avgt   50    39,828 ± 0,056  ns/op
BenchmarkSlice.equalsObject   32779  avgt   50  6284,251 ± 3,951  ns/op
```

After `equalsObject`
```
Benchmark                    (size)  Mode  Cnt     Score    Error  Units
BenchmarkSlice.equalsObject       1  avgt   50     4,947 ±  0,189  ns/op
BenchmarkSlice.equalsObject       7  avgt   50    11,924 ±  0,114  ns/op
BenchmarkSlice.equalsObject       8  avgt   50     5,180 ±  0,064  ns/op
BenchmarkSlice.equalsObject      16  avgt   50     6,468 ±  0,112  ns/op
BenchmarkSlice.equalsObject      32  avgt   50     9,759 ±  0,075  ns/op
BenchmarkSlice.equalsObject      64  avgt   50    13,810 ±  0,070  ns/op
BenchmarkSlice.equalsObject     127  avgt   50    27,591 ±  0,059  ns/op
BenchmarkSlice.equalsObject   32779  avgt   50  3829,908 ± 10,580  ns/op
```

### CODE REVIEW

**Comment by fmeiser** [comment]
@martint Another small improvement in performance.

File: src/test/java/io/airlift/slice/BenchmarkSlice.java [hunk 1]
```diff
@@ -44,11 +48,18 @@ public Object compareTo(BenchmarkData data)
...
+    {
+        return data.slice1.equalsUnchecked(0, data.slice2, 0, data.slice1.length());
+    }
+
     @State(Scope.Thread)
     public static class BenchmarkData
     {
         @Param({"1", "7", "8", "16", "32", "64", "127", "32779"})
-        private int size;
+        private int size = 1;
```
Comment (martint) [inline src/test/java/io/airlift/slice/BenchmarkSlice.java]: Why is this needed?

Comment (fmeiser) [inline src/test/java/io/airlift/slice/BenchmarkSlice.java]: Inital call of test method in `main` (line 95) fails because size is 0 in this case.

**Comment by martint** [comment]
Merged, thanks!

**Comment by fmeiser** [comment]
That was fast, thanks!
